src/file_pipe_log/file_mgr.rs

```diff
@@ -0,0 +1,604 @@
+    /// `Config.enable-log-recycle` is true.
+    pub fn initialize(&mut self) -> Result<()> {
+        let now = Instant::now();
+        let stale_capacity = std::cmp::min(
+            self.capacity.saturating_sub(self.active_files.read().len()),
+            max_stale_log_count(),
+        );
+        // If `stale_capacity` > 0, we should prepare stale files for later
+        // log recycling in advance.
+        if stale_capacity > 0 {
+            // Concurrent prepraring will bring more time consumption on racing. So, we just
+            // introduce a serial processing for preparing progress.
+            let mut stale_files = self.stale_files.write();
+            let (prepare_first_seq, prepare_stale_files_count) = (
+                stale_files.span().1 + 1,
+                stale_capacity.saturating_sub(stale_files.len()),
+            );
+            for seq in prepare_first_seq..prepare_first_seq + prepare_stale_files_count as FileSeq {
+                let file_id = FileId {
+                    queue: self.queue,
+                    seq,
+                };
+                stale_files.push_back(
+                    seq,
+                    self.build_stale_file(
+                        file_id,
+                        LogFileFormat::new(Version::V2, 0 /* alignment */),
+                    )?,
+                );
+            }
+        }
+        info!(
+            "LogQueue: {:?} preparing stale raft logs takes {:?}, prepared count: {}",
+            self.queue,
+            now.elapsed(),
+            self.stale_files.read().len(),
+        );
+        Ok(())
+    }
+
+    /// Generates a Fake log used for recycling.
+    ///
+    /// Attention, this function is only called when `Config.enable-log-recycle`
+    /// is true.
+    fn build_stale_file(
+        &self,
+        file_id: FileId,
+        format: LogFileFormat,
+    ) -> Result<FileWithFormat<F>> {
+        let path_id = FileCollection::<F>::get_valid_path(&self.paths, self.target_file_size);
+        let file_path = file_id.build_stale_file_path(&self.paths[path_id]);
+        let fd = Arc::new(self.file_system.create(&file_path)?);
+        let mut writer = self.file_system.new_writer(fd.clone())?;
+        let mut written = 0_usize;
+        let buf = vec![0; std::cmp::min(LOG_STALE_FILE_BUF_SIZE, self.target_file_size)];
+        while written <= self.target_file_size {
+            writer.write_all(&buf).unwrap_or_else(|e| {
+                panic!("failed to prepare stale file: {}", e);
+            });
+            written += buf.len();
+        }
+        // Metadata of stale files are not what we're truely concerned. So,
+        // they can be ignored by clear them here.
+        Ok(FileWithFormat {
+            handle: fd,
+            format,
+            path_id,
+        })
+    }
+
+    /// Returns a valid path for dumping new files.
+    #[inline]
+    pub fn get_valid_path(paths: &Paths, _target_file_size: usize) -> PathId {
```

Comment:
remove this function. move the assertion to ctor.

Reply:
IMO, It's better to keep it. 
This function is a basement for later `multi-dir` feature, which is expected to return a valid `Path` to user according to whether the free space in one `Path` is enough for the `target_file_size`.